contracts/axelar-gateway/src/auth.rs

```diff
@@ -0,0 +1,211 @@
+use axelar_soroban_interfaces::types::{ProofSignature, ProofSigner, WeightedSigner};
+use soroban_sdk::xdr::ToXdr;
+use soroban_sdk::{crypto::Hash, panic_with_error, Bytes, BytesN, Env, Vec};
+
+use crate::storage_types::AuthDataKey;
+use crate::{error::AuthError, event, auth};
+use axelar_soroban_interfaces::types::{Proof, WeightedSigners};
+
+pub(crate) fn init_auth_verifier(
+    env: Env,
+    domain_separator: BytesN<32>,
+    minimum_rotation_delay: u64,
+    previous_signer_retention: u64,
+    initial_signers: Vec<WeightedSigners>,
+) {
+    if env.storage().instance().has(&AuthDataKey::Initialized) {
+        panic!("Already initialized");
+    }
+
+    env.storage().instance().set(&AuthDataKey::Initialized, &true);
+
+    env.storage().instance().set(&AuthDataKey::Epoch, &0_u64);
+
+    // TODO: Do we need to manually expose these in a query, or can it be read directly off of storage in Stellar?
+    env.storage().instance().set(
+        &AuthDataKey::PreviousSignerRetention,
+        &previous_signer_retention,
+    );
+
+    env.storage()
+        .instance()
+        .set(&AuthDataKey::DomainSeparator, &domain_separator);
+
+    env.storage()
+        .instance()
+        .set(&AuthDataKey::MinimumRotationDelay, &minimum_rotation_delay);
+
+    if initial_signers.is_empty() {
+        panic_with_error!(env, AuthError::InvalidSigners);
+    }
+
+    for signers in initial_signers.into_iter() {
+        auth::rotate_signers_set(&env, signers, false);
+    }
+}
+
+pub(crate) fn validate_proof(env: &Env, data_hash: BytesN<32>, proof: Proof) -> bool {
+    let signer_set = proof.weighted_signers();
+
+    let signer_hash: BytesN<32> = env.crypto().keccak256(&signer_set.to_xdr(&env)).into();
+
+    let signer_epoch: u64 = env
+        .storage()
+        .persistent()
+        .get(&AuthDataKey::EpochBySignerHash(signer_hash.clone()))
+        .unwrap_or(0);
+
+    if signer_epoch == 0 {
+        panic_with_error!(env, AuthError::InvalidSigners);
+    }
+
+    let current_epoch: u64 = auth::epoch(&env);
+
+    let is_latest_signers: bool = signer_epoch == current_epoch;
+
+    let previous_signer_retention: u64 = env
+        .storage()
+        .instance()
+        .get(&AuthDataKey::PreviousSignerRetention)
+        .unwrap();
+
+    if current_epoch - signer_epoch > previous_signer_retention {
+        panic_with_error!(env, AuthError::InvalidSigners);
+    }
+
+    let msg_hash = auth::message_hash_to_sign(&env, signer_hash, data_hash);
+
+    if !auth::validate_signatures(&env, msg_hash, proof.clone()) {
+        panic_with_error!(env, AuthError::InvalidSignatures);
+    }
+
+    is_latest_signers
+}
+
+pub(crate) fn rotate_signers_set(env: &Env, new_signers: WeightedSigners, enforce_rotation_delay: bool) {
```

Comment:
```suggestion
pub(crate) fn rotate_signers(env: &Env, new_signers: WeightedSigners, enforce_rotation_delay: bool) {
```
this can be the same name now

Reply:
updated.